test/permuter/PermuterTest.cpp

```diff
@@ -0,0 +1,638 @@
+  EXPECT_EQ(internal->front().ops.front().op.metadata.write_sector,
+      write_sector);
+  EXPECT_EQ(internal->front().ops.front().op.metadata.bi_rw,
+      HWM_FLUSH_FLAG | HWM_WRITE_FLAG);
+
+  EXPECT_EQ(internal->back().num_meta, 0);
+  EXPECT_EQ(internal->back().checkpoint_epoch, 0);
+  EXPECT_FALSE(internal->back().has_barrier);
+  EXPECT_FALSE(internal->back().overlaps);
+  EXPECT_EQ(internal->back().ops.size(), 1);
+  EXPECT_EQ(internal->back().ops.front().abs_index, 1);
+  EXPECT_EQ(internal->back().ops.front().op.metadata.size, write_size);
+  EXPECT_EQ(internal->back().ops.front().op.metadata.write_sector,
+      write_sector);
+  EXPECT_EQ(internal->back().ops.front().op.metadata.bi_rw, HWM_WRITE_FLAG);
+}
+
+TEST(Permuter, InitDataVectorNoSplitFlush) {
+  vector<disk_write> test_epoch;
+
+  // Create a Checkpoint in the log since all logs start with one.
+  disk_write checkpoint;
+  checkpoint.metadata.write_sector = 0;
+  checkpoint.metadata.bi_flags = HWM_CHECKPOINT_FLAG;
+  checkpoint.metadata.bi_rw = HWM_CHECKPOINT_FLAG;
+  checkpoint.metadata.size = 0;
+  checkpoint.metadata.time_ns = 0;
+  test_epoch.push_back(checkpoint);
+
+  disk_write barrier;
+  barrier.metadata.bi_rw = HWM_FLUSH_FLAG | HWM_FUA_FLAG | HWM_WRITE_FLAG;
+  barrier.metadata.write_sector = 512;
+  barrier.metadata.size = 1024;
+  test_epoch.push_back(barrier);
+
+  TestPermuter tp;
+  tp.InitDataVector(test_epoch);
+  vector<epoch> *internal = tp.GetInternalEpochs();
+
+  EXPECT_EQ(internal->size(), 1);
+
+  EXPECT_EQ(internal->front().num_meta, 0);
+  EXPECT_EQ(internal->front().checkpoint_epoch, 0);
+  EXPECT_TRUE(internal->front().has_barrier);
+  EXPECT_FALSE(internal->front().overlaps);
+  EXPECT_EQ(internal->front().ops.size(), 1);
+
+  VerifyEpoch(internal->front(), test_epoch.begin() + 1, test_epoch.end(), 1);
+}
+
+TEST(Permuter, InitDataVectorNoSplitFlush2) {
+  vector<disk_write> test_epoch;
+
+  // Create a Checkpoint in the log since all logs start with one.
+  disk_write checkpoint;
+  checkpoint.metadata.write_sector = 0;
+  checkpoint.metadata.bi_flags = HWM_CHECKPOINT_FLAG;
+  checkpoint.metadata.bi_rw = HWM_CHECKPOINT_FLAG;
+  checkpoint.metadata.size = 0;
+  checkpoint.metadata.time_ns = 0;
+  test_epoch.push_back(checkpoint);
+
+  disk_write barrier;
+  barrier.metadata.bi_rw = HWM_FLUSH_FLAG | HWM_WRITE_FLAG;
+  barrier.metadata.write_sector = 512;
+  barrier.metadata.size = 0;
+  test_epoch.push_back(barrier);
+
+  TestPermuter tp;
+  tp.InitDataVector(test_epoch);
+  vector<epoch> *internal = tp.GetInternalEpochs();
+
+  EXPECT_EQ(internal->size(), 1);
+
+  EXPECT_EQ(internal->front().num_meta, 0);
+  EXPECT_EQ(internal->front().checkpoint_epoch, 0);
+  EXPECT_TRUE(internal->front().has_barrier);
+  EXPECT_FALSE(internal->front().overlaps);
+  EXPECT_EQ(internal->front().ops.size(), 1);
+
+  VerifyEpoch(internal->front(), test_epoch.begin() + 1, test_epoch.end(), 1);
+}
+
+TEST(Permuter, InitDataVectorCheckpointSkip) {
```

Comment:
Short description of what each test is for would be helpful

Reply:
done